File: Weekly Contest/Weekly Contest 170/Get_Watched_Videos_by_Your_Friends.java
//import java.util.Map.Entry;
class Solution {
    public List<String> watchedVideosByFriends(List<List<String>> watchedVideos, int[][] friends, int id, int level) {
        int[] lev=new int[friends.length];
        int index=1;
        for(int i=0; i<lev.length; i++) lev[i]=-1;
        Queue<Integer> qu = new LinkedList<Integer>();
        lev[id]=0;
        for(int i=0; i<friends[id].length; i++){
            int f=friends[id][i];
            qu.offer(f);
            lev[f]=index;
        }
        while(qu.size()>0){
            if(index==level) break;
            int len=qu.size();
            ++index;
            while(len-->0){
                int v=qu.poll();
                for(int i=0; i<friends[v].length; i++){
                    int f=friends[v][i];
                    if(lev[f]==-1){
                        lev[f]=index; 
                        qu.offer(f);
                    }
                }
            }
        }
        Map<String, Integer> map = new HashMap<String, Integer>();
        for(int i=0; i<friends.length; i++){
            if(lev[i]==level){
                List<String> tmp=watchedVideos.get(i);
                for(int j=0; j<tmp.size(); j++){
                    int v=0;
                    if(map.containsKey(tmp.get(j))) v=map.get(tmp.get(j));
                    ++v;
                    map.put(tmp.get(j), v);
                }
            }
        }
		    List<Map.Entry<String,Integer>> list = new ArrayList<Map.Entry<String,Integer>>(map.entrySet());
        Collections.sort(list,new Comparator<Map.Entry<String,Integer>>() {
        public int compare(Map.Entry<String, Integer> o1, Map.Entry<String, Integer> o2) {
            if(o1.getValue()==o2.getValue()) return o1.getKey().compareTo(o2.getKey());
            return o1.getValue() - o2.getValue();
        }
        });
        List<String> result = new ArrayList<String>();
        for(Map.Entry<String, Integer> mapping:list){ 
            result.add(mapping.getKey());
        } 
        return result;
    }
}

File: Weekly Contest/Weekly Contest 170/Get_Watched_Videos_by_Your_Friends.cpp
class Solution {
public:
    typedef pair<string, int> PAIR;
    static bool cmp_by_value(const PAIR& lhs, const PAIR& rhs) {
        if(lhs.second==rhs.second){
            if(lhs.first.compare(rhs.first)>0) return false;
            return true;
        }
        return lhs.second < rhs.second;
    }
    vector<string> watchedVideosByFriends(vector<vector<string>>& watchedVideos, vector<vector<int>>& friends, int id, int level) {
        
        int len=friends.size();       
        int lev[128];
        int index=1;
        for(int i=0; i<len; i++) lev[i]=-1;
        queue<int> qu;
        lev[id]=0;
        for(int i=0; i<friends[id].size(); i++){
            qu.push(friends[id][i]);
            lev[friends[id][i]]=index;
        }
        while(!qu.empty()){
            if(index==level) break;
            int l=qu.size();
            ++index;
            while(l-->0){
                int v=qu.front();
                qu.pop();
                for(int i=0; i<friends[v].size(); i++){
                    if(lev[friends[v][i]]==-1){
                        qu.push(friends[v][i]);
                        lev[friends[v][i]]=index;
                    }
                }
            }
        }
        unordered_map<string, int> mmap;
        for(int i=0; i<len; i++){
            if(lev[i]==level){
                vector<string> tmp=watchedVideos[i];
                for(int j=0; j<tmp.size(); j++){
                    if(mmap.find(tmp[j])!=mmap.end()) ++mmap[tmp[j]];
                    else mmap[tmp[j]]=1;
                }
            }
        }
        
        vector<string> result;
        vector<PAIR> temp(mmap.begin(), mmap.end());
        sort(temp.begin(), temp.end(), cmp_by_value);
        for(int i=0; i<temp.size(); i++) result.push_back(temp[i].first);
        return result;
    }
};
